src/Microsoft.AspNetCore.Blazor.Browser.JS/src/Interop/InvokeDotNetMethodWithJsonMarshalling.ts

```diff
@@ -0,0 +1,183 @@
+import { platform } from '../Environment';
+import { System_String, Pointer } from '../Platform/Platform';
+import { getRegisteredFunction } from './RegisteredFunction';
+import { error } from 'util';
+
+export interface MethodOptions {
+  type: TypeInstance;
+  method: MethodInstance;
+}
+
+// Keep in sync with InvocationResult.cs
+export interface InvocationResult {
+  succeeded: boolean;
+  result?: any;
+  message?: string;
+}
+
+export interface MethodInstance {
+  name: string;
+  typeArguments?: { [key: string]: TypeInstance }
+  parameterTypes?: TypeInstance[];
+}
+
+export interface TypeInstance {
+  assembly: string;
+  name: string;
+  typeArguments?: { [key: string]: TypeInstance };
+}
+
+export function invokeDotNetMethod<T>(methodOptions: MethodOptions, ...args: any[]): (T | null) {
+  return invokeDotNetMethodCore(methodOptions, null, ...args);
+}
+
+let registrations = {};
+
+function resolveRegistration(methodOptions: MethodOptions) {
+  let existing = registrations[methodOptions.type.assembly];
+  existing = existing && registrations[methodOptions.type.name];
+  existing = existing && registrations[methodOptions.method.name];
+  if (existing !== undefined) {
+    return existing;
+  } else {
+    const method = platform.findMethod(
+      'Microsoft.AspNetCore.Blazor.Browser',
+      'Microsoft.AspNetCore.Blazor.Browser.Interop',
+      'JavaScriptInvoke',
+      'FindDotNetMethod');
+
+    const serializedOptions = platform.toDotNetString(JSON.stringify(methodOptions));
+    const result = platform.callMethod(method, null, [serializedOptions]);
+    const registration = platform.toJavaScriptString(result as System_String);
+
+    if (registrations[methodOptions.type.assembly] === undefined) {
```

Comment:
Not a big deal, but it's more idiomatic to write `if (!(methodOptions.type.assembly in registrations))` or just `if (!registrations[methodOptions.type.assembly])` since we know that its values aren't falsey.